### PR TITLE
Release metal3 0.6.0

### PR DESCRIPTION
**Pull request by Kristian-ZH**
This release includes the following changes:

- The Ironic images are rebased on the `2023.2` dependencies
- The Ironic images are now Aligned with the upstream version
- The BMO CRDs are aligned with the upstream
- The VMedia TLS is now configurable
- The Predictable NIC naming convention is now configurable

### CODE REVIEW

File: packages/ironic/charts/templates/certificates.yaml [hunk 1]
```diff
@@ -38,6 +38,7 @@ spec:
     kind: Issuer
     name: ca-issuer
   secretName: ironic-inspector-cert
+{{- if .Values.global.enable_vmedia_tls -}}
```
Comment (hardys) [inline packages/ironic/charts/templates/certificates.yaml:41]: I think the whitespace control here is wrong - in my environment it ends up like `secretName: ironic-inspector-cert---` - I think removing one of the `-` characters in this expression will fix it

Comment (Kristian-ZH) [reply packages/ironic/charts/templates/certificates.yaml]: Fixed, thanks!

**Comment by ipetrov117** [comment]
I will do a deployment on Sylva later today and update on whether it works there as well.

**Review by ipetrov117 — CHANGES_REQUESTED** [review]
Tried deploying the changes on Sylva's side and the cluster failed to deploy. 

The reason for the failure is that we have removed the `clusterctl.cluster.x-k8s.io: ""` label from all our BMO CRDs. This label is needed for [pivoting](https://github.com/metal3-io/cluster-api-provider-metal3?tab=readme-ov-file#pivoting-%EF%B8%8F) BMO resources from the `bootstrap` cluster to the `management` cluster.

I suggest bringing back these labels. If there is a specific reason for removing them, we would need to have further discussions on how to handle this scenario.

File: packages/baremetal-operator/charts/crds/customresource-baremetalhosts.yaml [hunk 2]
```diff
@@ -3,10 +3,7 @@ apiVersion: apiextensions.k8s.io/v1
 kind: CustomResourceDefinition
 metadata:
   annotations:
-    cert-manager.io/inject-ca-from: baremetal-operator-system/baremetal-operator-serving-cert
     controller-gen.kubebuilder.io/version: v0.12.1
-  labels:
-    clusterctl.cluster.x-k8s.io: ""
```
[the remark this answers is not in the document]
Comment (hardys) [reply packages/baremetal-operator/charts/crds/customresource-baremetalhosts.yaml:9]: Yes see the [CAPM3 docs](https://github.com/metal3-io/cluster-api-provider-metal3#pivoting-%EF%B8%8F) for more information.

Also is the removal of the cert-manager annotation intentional or just a consequence of copying the CRDs?

[the remark this answers is not in the document]
Comment (Kristian-ZH) [reply packages/baremetal-operator/charts/crds/customresource-baremetalhosts.yaml:9]: consequence of copying the CRDs

[the remark this answers is not in the document]
Comment (Kristian-ZH) [reply packages/baremetal-operator/charts/crds/customresource-baremetalhosts.yaml:9]: Will revert these changes

**Comment by hardys** [comment]
@Kristian-ZH since this release changes the CRDs we need to decide how to handle that, these changes should be backwards compatible so I think our options are either document manually applying the crds subdir prior to helm upgrade, or consider creating a new `-crds` chart.

@ipetrov117 are you aware of any specific handling of this scenario in Sylva for other charts?

**Comment by Kristian-ZH** [comment]
> I think our options are either document manually applying the crds subdir prior to helm upgrade, or consider creating a new -crds chart.

I was thinking to document that and to leave the decision for extracting crds to a dedicated dir for later because this should be aligned with all the charts, not only Metal3

**Comment by ipetrov117** [comment]
> @ipetrov117 are you aware of any specific handling of this scenario in Sylva for other charts?

@hardys, I am not aware of any specific convention that is being followed. AFAIK Sylva use whatever the main chart provides in terms of deployment logic. For instance, if the chart `foo` uses 2 charts for its deployment logic (`foo-crds` and `foo`), then 2 units are created in Sylva  - `foo-crds` and `foo`,  where `foo` depends on `foo-crds`. Example can be found [here](https://gitlab.com/sylva-projects/sylva-core/-/blob/main/charts/sylva-units/values.yaml?ref_type=heads#L2449).

My 2 cents on the metal3 chart upgrade discussion.. There doesn't seem to be a concrete approach when it comes to upgrading CRDs. Looking at the helm [doc](https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#install-a-crd-declaration-before-using-the-resource) regarding this, we have 2 possible approaches:

1. Have a `crds` directory in the chart and leave helm to deploy the crds for us. This is the current approach that we and [Sylva](https://gitlab.com/sylva-projects/sylva-elements/helm-charts/metal3) seem to be taking. While easier, this approach does not handle CRD upgrades well (at least from what I understand).
2. Have a `foo-crd` and `foo` charts, where `foo-crd` is always deployed before `foo`.

IMO documenting the manual CRD sync steps is an okay workaround as of this moment, but for the long term we should think about option (2). By having the CRDs in a chart outside of our main chart we can always ensure that they have been upgraded to the desired state before running the main chart upgrade.

File: packages/baremetal-operator/charts/crds/customresource-baremetalhosts.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (Kristian-ZH) [reply packages/baremetal-operator/charts/crds/customresource-baremetalhosts.yaml]: Addressed

**Review by ipetrov117 — APPROVED** [review]
Works as expected now. Sylva cluster deployed successfully 👍

**Comment by hardys** [comment]
@Kristian-ZH I tested in the metal3-demo environment and I think we have an issue with UEFI, we see the BMH fail to start inspection like:

```
  Normal  InspectionError    54m   metal3-baremetal-controller  Failed to inspect hardware. Reason: unable to start inspection: Failed to download image http://192.168.125.10:6180/uefi_esp.img, reason: HTTPConnectionPool(host='192.168.125.10', port=6180): Max retries exceeded with url: /uefi_esp.img (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f5602727c50>: Failed to establish a new connection: [Errno 113] EHOSTUNREACH'))
```

Looking at the ironic config I think we can see the problem:
```
shardy@localhost:~> kubectl exec -n metal3-system metal3-metal3-ironic-677bc5c8cc-8cqdg -c ironic -- cat /etc/ironic/ironic.conf | grep uefi_esp
bootloader = http://192.168.125.10:6180/uefi_esp.img
```

```
shardy@localhost:~> kubectl get service -n metal3-system
NAME                                                    TYPE           CLUSTER-IP     EXTERNAL-IP      PORT(S)                                        AGE
baremetal-operator-controller-manager-metrics-service   ClusterIP      10.43.95.163   <none>           8443/TCP                                       86m
baremetal-operator-webhook-service                      ClusterIP      10.43.210.62   <none>           443/TCP                                        86m
metal3-mariadb                                          ClusterIP      10.43.136.41   <none>           3306/TCP                                       86m
metal3-metal3-ironic                                    LoadBalancer   10.43.57.230   192.168.125.10   6185:31785/TCP,5050:30890/TCP,6385:30479/TCP   86m
```

**Comment by hardys** [comment]
Additional note, in the previous ironic image we had

```
bootloader = {{ env.IRONIC_BOOT_BASE_URL }}/uefi_esp.img
```

But I think due to the upstream rebase that got lost and is now e.g

```
bootloader = http://{{ env.IRONIC_URL_HOST }}:{{ env.HTTP_PORT }}/uefi_esp.img
```

**Review by alknopfler — APPROVED** [review]
LGTM.
Tested on the lab with the next configuration:
```
global:
  ironicIP: "10.168.200.22"
  enable_vmedia_tls: false

# See /opt/media for media-server images
metal3-media:
  service:
    type: NodePort
    port: 6280

metal3-ironic:
  global:
    predictableNicNames: "false"
  service:
    type: NodePort
  persistence:
    ironic:
      storageClass: local-path

metal3-mariadb:
  persistence:
    storageClass: local-path

metal3-media:
  mediaVolume:
    hostPath: /home/metal3/bmh-image-cache
```
Cluster HA provisioned and metallb working fine as expected:
- TLS not configured on the v-media
- ethX interfaces configured on the manifests and working fine

**Comment by hardys** [comment]
I re-tested after https://build.opensuse.org/request/show/1144913 merged and all works OK now in the metal3-demo environment

**Review by hardys — APPROVED** [review]

**Pull request merged**